File: benchmarks/c/array_sum.c
#include <stdio.h>

#define SIZE 1000
#define ITERATIONS 10000

static inline void black_box_ptr(const void *ptr) {
    __asm__ __volatile__("" : : "r"(ptr) : "memory");
}

int main(void) {
    long arr[SIZE];
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }

    long total = 0;
    for (int iter = 0; iter < ITERATIONS; iter++) {
        black_box_ptr(arr);
        long sum = 0;
        for (int i = 0; i < SIZE; i++) {
            sum += arr[i];
        }
        total += sum;
    }

    printf("Completed %d iterations, total: %ld\n", ITERATIONS, total);
    return 0;
}

File: benchmarks/ocaml/array_sum.ml
let () =
  let size = 1000 in
  let iterations = 10_000 in
  let arr = Array.init size Fun.id in
  let total = ref 0 in
  for _iter = 0 to iterations - 1 do
    let sum = ref 0 in
    for i = 0 to size - 1 do
      sum := !sum + arr.(i)
    done;
    total := !total + !sum
  done;
  Printf.printf "Completed %d iterations, total: %d\n" iterations !total
